### ISSUE TITLE
Incremental STP Solving

### ISSUE THREAD

**Issue by sreeshmaheshwar** [stp/stp#483]
Does STP support incremental SMT solving techniques (e.g. reusing learned lemmas) as other solvers, like Z3 and Boolector, do? (Can't find anything in the docs about how to enable an incremental solver and haven't observed any performance differences when using `push` and `pop` in an incremental way with STP)

If not, are there any plans for this?

**Comment by TrevorHansen** [comment]
You're right that STP doesn't keep learned clauses between invocations to the SAT solver.

It has some simple analysis, for example if a problem is unsatisfiable, and you push more frames onto it, STP knows that those problems are unsatisfiable, too.

There's lots we could do to make incremental solving faster. If you provide us with some problems that are really slow with STP ,compared to other solvers, I'll see if there's anything we can do to speed up STP's solving of them.

**Comment by sreeshmaheshwar** [comment]
Many apologies for the late reply. Thanks @TrevorHansen for your response and willingness.

Attached are some example files, using incremental querying with `push`/`pop`, for which Z3 outperforms STP, for me at least. They are KLEE generated, so sorry for the unreadability (I did try to condense them and remove redundancy but some remains) - hopefully it is not so bad that it isn't useful.

The time differences for me on my Ubuntu 20.04 machine are far more dramatic (1.5 minutes for STP vs 1.5 seconds for Z3) though I suppose the configuration is outdated. On my Ubuntu 22.04 machine, the differences are 15.1s (STP) vs 0.7s (Z3) and 8.7s (STP) vs 0.3s (Z3). I use mainline STP throughout, with Z3 4.8.12 (22.04) and 4.8.7 (20.04).

I can produce more files that show STP being outperformed in such an incremental setting, so do let me know if that would be helpful, though my current process of printing into SMT-LIB is a bit painful. If these differences are not substantial enough and you'd like for a small number of queries that produce larger differences, I can try for that as well but I've struggled to find this so far (through KLEE).

Please let me know if there's anything else that I can help out with on this topic! I'd be very happy to see STP supporting incremental solving.

SMT-LIB2 files: [dircolors.txt](https://github.com/stp/stp/files/15456941/dircolors.txt), [b64.txt](https://github.com/stp/stp/files/15456459/b64.txt)

**Comment by sreeshmaheshwar** [comment]
@TrevorHansen just wanted to friendly ping on this as I have some free time now - are these examples sufficient?

**Comment by sreeshmaheshwar** [comment]
Thanks for the response @TrevorHansen. 

Indeed, Z3's incremental solver that is used for these benchmarks performs worse on harder problems due to lack of pre-processing (the term *simplifications* is overloaded in the context of Z3). See the examples linked in the relevant discussion https://github.com/Z3Prover/z3/discussions/7203 where, as a result, I was confused by the outperformance of Z3's incremental solver when used non-incrementally in mainline KLEE. It also suggests that such hard problems are rarer in KLEE - though I will keep an eye out for slower queries during longer runs.

The two benchmarks I provided in our current discussion are from my experimental fork of KLEE. This incremental work is experimental and concludes shortly, so I should clarify that there is no concrete need for STP incrementality yet (from KLEE), in the case that it is difficult to implement or not worth it in your opinion. As an aside, you mention an incremental solver that switches to a non-incremental one after a timeout - this is supported by Z3.

Speaking of mainline KLEE, your suggestion of timeout-based _portfolio solving_ using Z3's incremental (but used non-incrementally) solver and STP's (normal) solver seemed to me to be a near-optimal KLEE configuration. Given that STP's un-simplified solving (that I didn't know was possible) now seems to be worse than Z3's (but I will try to confirm this on normal KLEE workloads), and STP's normal solving, from mine and prior experiments, is faster than Z3's (though this may change if only expensive queries are considered), this might be the way forwards. Thanks for the suggestion/confirmation.

In any case, for a less intrusive change that does not combine solvers, does STP support running an un-simplified solver (that does not support incrementality, just `disable-simplifications`) but defaulting back to a normal one after a specified timeout? I'd have guessed that this would still lead to KLEE improvements, though the 16s -> 11s difference you highlighted was less than we observed for Z3 - but STP's normal solver would've been faster than Z3's so I suppose it's not unsurprising. Of course, KLEE could do this itself by constructing the solver and AST from scratch, but wondering if STP has existing support for this.

**Comment by sreeshmaheshwar** [comment]
Oops, just realised you'd already suggested this in that discussion - has this been implemented then?

**Comment by aytey** [comment]
Incremental solving is implemented and in master, and it is on by default — no flag needed. A session becomes incremental at its first `push`: one SAT solver and one bit-blasted encoding stay alive across `(check-sat)` calls, with retractable levels asserted as SAT assumptions, so a `pop` retracts a level by simply no longer assuming it and learned clauses survive by construction. `check-sat-assuming` and `get-unsat-assumptions` came with it. There is a full write-up in `docs/incremental-solving.rst`.

@sreeshmaheshwar — thank you for the two files, they were exactly what was needed, and both are now well inside the range you were after.

Median of three sequential runs on an idle 24-core box, STP at 377ad4fc with the CaDiCaL backend. "Before" is d1c9293e, the commit immediately preceding the incremental series (#832–#836), so the difference is attributable to this work rather than to everything else that has landed since you filed:

| | `dircolors` (5000 `check-sat`, 5833 `push`) | `b64` (1000 `check-sat`, 44165 `assert`) |
| --- | --- | --- |
| before incremental | 34.23 s | 17.26 s |
| today, default | **1.07 s** | **1.11 s** |
| today, forced with `--incremental` | **1.05 s** | **1.08 s** |
| speedup | **32×** | **16×** |

Peak RSS moves with it: 123 MB → 17 MB on `dircolors`, 80 MB → 31 MB on `b64`.

The default and the forced column differ by about 3%, which is the whole cost of the automatic policy on these two files. Neither has a `set-logic`, so the theory-neutral threshold applies and the driver engages at the third solve — two batch solves out of 5000 and 1000. Forcing it earlier is not worth a flag here.

The answers are unchanged: all 5000 and all 1000 verdicts are identical to the ones the batch pipeline produces, and identical again to the pre-incremental binary above.

Worth being straight about the comparison you originally drew: these are not the z3 numbers you reported (0.7 s and 0.3 s), and that was your machine rather than this one, so the two sets are not directly comparable in either direction. The gap that motivated this issue — a minute and a half against a second and a half — is gone, but I would not claim parity from measurements taken on different hardware.

If you still have the KLEE setup, we would be glad to know whether this makes things better on your side, and more files of the kind you offered are very welcome — particularly any where it does *not* help, since those are the interesting ones now.

Closing this as implemented.
